File: tools/flatten_validation/flatten.py
#!/usr/bin/env python3
#
# Copyright 2023 Brian T. Park
#
# MIT License

"""
Flatten the `validation_data.json` file on the STDIN to a more compact
TXT file format on the STDOUT. The TXT file can be more than 10X smaller than
the JSON file.

Usage:
$ flatten.py < validation_data.json > validation_data.txt
"""

from typing import List, Dict
import argparse
import logging
import sys
import json

from acetimetools.data_types.validation_types import (
    TestItem,
    TestEntry,
    ValidationData
)


def main() -> None:
    # Configure command line flags.
    parser = argparse.ArgumentParser(
        description='Flatten validation_data.json'
    )

    # DST blacklist JSON file.
    parser.add_argument(
        '--blacklist',
        type=str,
        help='JSON file containing DST blacklist',
    )

    # Ignore blacklist. Useful for debugging 3rd party timezones which have
    # inconsistencies with AceTime (or Hinnant date).
    parser.add_argument(
        '--ignore_blacklist',
        action='store_true',
        help='Ignore the --blacklist flag, useful for debugging',
    )

    # Parse the command line arguments
    args = parser.parse_args()

    # Configure logging. This should normally be executed after the
    # parser.parse_args() because it allows us set the logging.level using a
    # flag.
    logging.basicConfig(level=logging.INFO)

    # Read the JSON on the STDIN
    validation_data = json.load(sys.stdin)

    # Read the DST blacklist file if given.
    if args.blacklist and not args.ignore_blacklist:
        with open(args.blacklist) as f:
            blacklist = json.load(f)
    else:
        blacklist = {}

    # Generate the validation_*.{h, cpp} files
    flattener = Flattener(blacklist=blacklist)
    flattener.flatten(validation_data)


class Flattener:
    """Convert JSON into flat file of the form:
    HEADER
    start_year: {int}
    until_year: {int}
    epoch_year: {int}
    ...

    ZONE {zone}
    TRANSITIONS {len}
    {# epoch offset dst y M d h m s abbrev type}
    ...
    SAMPLES {len}
    {epoch offset dst y M d h m s abbrev type}
    ...

    ZONE {zone}
    ...
    """

    def __init__(self, blacklist: Dict[str, str]):
        self.blacklist = blacklist

    def flatten(self, validation_data: ValidationData):
        self.print_header(validation_data)
        test_data = validation_data['test_data']
        for zone, test_entry in test_data.items():
            self.print_zone_data(zone, test_entry)

    def print_header(self, validation_data: ValidationData) -> None:
        print('HEADER')
        print('start_year', validation_data['start_year'])
        print('until_year', validation_data['until_year'])
        print('epoch_year', validation_data['epoch_year'])
        print('has_valid_abbrev', validation_data['has_valid_abbrev'])
        print('has_valid_dst', validation_data['has_valid_dst'])
        print()

    def print_zone_data(self, zone: str, entry: TestEntry) -> None:
        print(f'ZONE {zone}')

        transitions = entry['transitions']
        print(f'TRANSITIONS {len(transitions)}')
        self.print_test_items(transitions)

        samples = entry['samples']
        print(f'SAMPLES {len(samples)}')
        self.print_test_items(samples)
        print()

    def print_test_items(self, items: List[TestItem]) -> None:
        if len(items):
            print("# line       epoch    utc    dst    y  m  d  h  m  s  \
abbrev type")
        line_number = 0
        for item in items:
            epoch_seconds = item['epoch']
            total_offset = item['total_offset']
            delta_offset = item['dst_offset']
            year = item['y']
            month = item['M']
            day = item['d']
            hour = item['h']
            minute = item['m']
            second = item['s']
            abbrev_value = item['abbrev']
            abbrev = f'{abbrev_value}' if abbrev_value else '-'
            type = item['type']

            print(f"{line_number:>6} \
{epoch_seconds:11} {total_offset:6} {delta_offset:6} \
{year:4} {month:2} {day:2} {hour:2} {minute:2} {second:2} {abbrev:>7} \
{type:>4}")
            line_number += 1


if __name__ == '__main__':
    main()

File: tools/flatten_validation/flatten.go
// Copyright 2023 Brian T. Park
//
// MIT License
//
// Read the validation JSON data on the stdin, and print out a flatten version
// that is easier to read by humans for debugging purposes.
//
// Usage:
// $ go run flatten.go < data.json > data.txt
//
// Performance:
// The Go version is actually 40-50% slower than the Python version. This is
// probably because the Python json package is implemented in C.

package main

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
)

//-----------------------------------------------------------------------------

// The collection of all zones and their test items.
type ValidationData struct {
	StartYear      int      `json:"start_year"`
	UntilYear      int      `json:"until_year"`
	EpochYear      int      `json:"epoch_year"`
	Source         string   `json:"source"`
	Version        string   `json:"version"`
	TzVersion      string   `json:"tz_version"`
	HasValidAbbrev bool     `json:"has_valid_abbrev"`
	HasValidDst    bool     `json:"has_valid_dst"`
	TData          TestData `json:"test_data"`
}

// A map of zoneName to an array of TestItems.
type TestData map[string]TestEntry

// A pair of transitions and samples.
type TestEntry struct {
	Transitions []TestItem `json:"transitions"`
	Samples     []TestItem `json:"samples"`
}

// A test item struct contains the epochSeconds with its local date time
// components.
type TestItem struct {
	EpochSeconds int64  `json:"epoch"`
	TotalOffset  int32  `json:"total_offset"`
	DstOffset    int32  `json:"dst_offset"`
	Year         int16  `json:"y"`
	Month        uint8  `json:"M"`
	Day          uint8  `json:"d"`
	Hour         uint8  `json:"h"`
	Minute       uint8  `json:"m"`
	Second       uint8  `json:"s"`
	Abbrev       string `json:"abbrev"`
	ItemType     string `json:"type"` // "A", "B", "a", "b", "S", "T"
}

//-----------------------------------------------------------------------------

func main() {
	data := readData()
	writeData(data)
}

func readData() *ValidationData {
	var data ValidationData
	err := json.NewDecoder(os.Stdin).Decode(&data)
	if err != nil {
		log.Fatal(err)
	}
	return &data
}

func writeData(data *ValidationData) {
	fmt.Println("HEADER")
	fmt.Println("start_year", data.StartYear)
	fmt.Println("until_year", data.UntilYear)
	fmt.Println("epoch_year", data.EpochYear)
	fmt.Println("has_valid_abbrev", data.HasValidAbbrev)
	fmt.Println("has_valid_dst", data.HasValidDst)
	fmt.Println()

	testData := data.TData
	for zone, entry := range testData {
		writeTestEntry(zone, &entry)
	}
}

func writeTestEntry(zone string, entry *TestEntry) {
	fmt.Println("ZONE", zone)

	transitions := entry.Transitions
	fmt.Println("TRANSITIONS", len(transitions))
	writeTestItems(transitions)

	samples := entry.Samples
	fmt.Println("SAMPLES", len(samples))
	writeTestItems(samples)
	fmt.Println()
}

func writeTestItems(items []TestItem) {
	if len(items) != 0 {
		fmt.Println(
			"# line       epoch    utc    dst    y  m  d  h  m  s  abbrev type")
	}

	for i := range items {
		item := &items[i]

		var abbrev string
		if item.Abbrev == "" {
			abbrev = "-"
		} else {
			abbrev = item.Abbrev
		}

		fmt.Printf(
			"%6d %11d %6d %6d %4d %2d %2d %2d %2d %2d %7s %4s\n",
			i,
			item.EpochSeconds,
			item.TotalOffset,
			item.DstOffset,
			item.Year,
			item.Month,
			item.Day,
			item.Hour,
			item.Minute,
			item.Second,
			abbrev,
			item.ItemType,
		)
	}
}
